src/Mod/Path/PathScripts/PathVcarve.py

```diff
@@ -0,0 +1,280 @@
+else:
+    PathLog.setLevel(PathLog.Level.INFO, PathLog.thisModule())
+
+
+# Qt tanslation handling
+def translate(context, text, disambig=None):
+    return QtCore.QCoreApplication.translate(context, text, disambig)
+
+
+class ObjectVcarve(PathEngraveBase.ObjectOp):
+    '''Proxy class for Vcarve operation.'''
+
+    def opFeatures(self, obj):
+        '''opFeatures(obj) ... return all standard features and edges based geomtries'''
+        return PathOp.FeatureTool | PathOp.FeatureHeights | PathOp.FeatureBaseFaces
+
+    def setupAdditionalProperties(self, obj):
+        if not hasattr(obj, 'BaseShapes'):
+            obj.addProperty("App::PropertyLinkList", "BaseShapes", "Path",
+                            QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                                "Additional base objects to be engraved"))
+        obj.setEditorMode('BaseShapes', 2)  # hide
+        if not hasattr(obj, 'BaseObject'):
+            obj.addProperty("App::PropertyLink", "BaseObject", "Path",
+                            QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                            "Additional base objects to be engraved"))
+        obj.setEditorMode('BaseObject', 2)  # hide
+
+    def initOperation(self, obj):
+        '''initOperation(obj) ... create vcarve specific properties.'''
+        obj.addProperty("App::PropertyFloat", "Discretize", "Path",
+                        QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                        "The deflection value for discretizing arcs"))
+        obj.addProperty("App::PropertyFloat", "Threshold", "Path",
+                        QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                        "cutoff for removing colinear segments (degrees). \
+                        default=10.0."))
+        obj.addProperty("App::PropertyFloat", "Tolerance", "Path",
+                QtCore.QT_TRANSLATE_NOOP("PathVcarve", ""))
+        obj.Threshold = 10.0
+        obj.Discretize = 0.01
+        obj.Tolerance = PathPreferences.defaultGeometryTolerance()
+        self.setupAdditionalProperties(obj)
+
+    def opOnDocumentRestored(self, obj):
+        # upgrade ...
+        self.setupAdditionalProperties(obj)
+
+    def buildPathMedial(self, obj, Faces):
+        '''constructs a medial axis path using openvoronoi'''
+
+        def insert_many_wires(vd, wires):
+            for wire in wires:
+                PathLog.debug('discretize value: {}'.format(obj.Discretize))
+                pts = wire.discretize(QuasiDeflection=obj.Discretize)
+                ptv = [FreeCAD.Vector(p[0], p[1]) for p in pts]
+                ptv.append(ptv[0])
+
+                for i in range(len(pts)):
+                    vd.addSegment(ptv[i], ptv[i+1])
+
+        def calculate_depth(MIC):
+            # given a maximum inscribed circle (MIC) and tool angle,
+            # return depth of cut.
+            maxdepth = obj.ToolController.Tool.CuttingEdgeHeight
+            toolangle = obj.ToolController.Tool.CuttingEdgeAngle
```

Comment:
Not all tools have these attributes - especially with the new `ToolBit`. For a v-bit the cutting edge height needs to be calculated from the `Diameter` and the `CuttingEdgeAngle`. OTOH,  an endmill does not have a `CuttingEdgeAngle` attribute, it is by definition 90°.

Maybe we should add some utility functions to `ToolBit` itself which returns values for some basic properties, like the ones above, regardless if they are an Attribute of the actual tool bit or not.

Reply:
Now checking for the needed attributes here but I think expanding ToolBit is a good idea